ext/MooncakeDynamicExpressionsExt.jl

```diff
@@ -0,0 +1,486 @@
+module MooncakeDynamicExpressionsExt
+
+using DynamicExpressions:
+    DynamicExpressions as DE,
+    AbstractExpressionNode,
+    Nullable,
+    constructorof,
+    branch_copy,
+    leaf_copy,
+    get_child,
+    get_children,
+    set_children!
+using Mooncake
+using Mooncake: NoTangent
+using Random: AbstractRNG
+
+################################################################################
+# Main tangent type
+################################################################################
+
+mutable struct TangentNode{Tv,D}
+    const degree::UInt8
+    val::Union{Tv,NoTangent}
+    children::NTuple{D,Union{@NamedTuple{null::NoTangent,x::TangentNode{Tv,D}},NoTangent}}
+end
+
+function TangentNode{Tv,D}(
+    val_tan::Union{Tv,NoTangent}, children::Vararg{Union{TangentNode{Tv,D},NoTangent},deg}
+) where {Tv,D,deg}
+    return TangentNode{Tv,D}(
+        UInt8(deg),
+        val_tan,
+        ntuple(i -> i <= deg ? _wrap_nullable(children[i]) : NoTangent(), Val(D)),
+    )
+end
+
+function Mooncake.tangent_type(::Type{<:AbstractExpressionNode{T,D}}) where {T,D}
+    Tv = Mooncake.tangent_type(T)
+    return Tv === NoTangent ? NoTangent : TangentNode{Tv,D}
+end
+function Mooncake.tangent_type(::Type{TangentNode{Tv,D}}) where {Tv,D}
+    return TangentNode{Tv,D}
```

Comment:
Not sure about this. Wouldn't this be used in higher-order derivatives, and therefore be incorrect?

(Should I make the TangentNode{Tv,D} <: AbstractExpressionNode{Tv,D} so that it can be used as its own higher order derivative? Tricky...)

Reply:
(Done. I think now in principle you should be able to do derivatives _OF_ the TangentNode)